typing/typemod.ml

```diff
@@ -1449,262 +1540,299 @@ and type_module_aux ~alias sttn funct_body anchor env smod =
-          cl;
-        let (classes, new_env) = Typeclass.class_type_declarations env cl in
-        Tstr_class_type
-          (List.map (fun cl ->
-               (cl.Typeclass.clsty_ty_id,
-                cl.Typeclass.clsty_id_loc,
-                cl.Typeclass.clsty_info)) classes),
-(*  TODO: check with Jacques why this is here
-           Tstr_type
-             (List.map (fun (_, _, i, d, _, _) -> (i, d)) classes) ::
-           Tstr_type
-             (List.map (fun (_, _, _, _, i, d) -> (i, d)) classes) :: *)
-        List.flatten
-          (map_rec
-             (fun rs decl ->
-                let open Typeclass in
-                [Sig_class_type(decl.clsty_ty_id, decl.clsty_ty_decl, rs);
-                 Sig_type(decl.clsty_obj_id, decl.clsty_obj_abbr, rs);
-                 Sig_type(decl.clsty_typesharp_id, decl.clsty_abbr, rs)])
-             classes []),
-        new_env
-    | Pstr_include sincl ->
-        let smodl = sincl.pincl_mod in
-        let modl =
-          Builtin_attributes.warning_scope sincl.pincl_attributes
-            (fun () -> type_module true funct_body None env smodl)
-        in
-        (* Rename all identifiers bound by this signature to avoid clashes *)
-        let sg = Subst.signature Subst.identity
-            (extract_sig_open env smodl.pmod_loc modl.mod_type) in
-        List.iter (check_sig_item names loc) sg;
-        let new_env = Env.add_signature sg env in
-        let incl =
-          { incl_mod = modl;
-            incl_type = sg;
-            incl_attributes = sincl.pincl_attributes;
-            incl_loc = sincl.pincl_loc;
-          }
-        in
-        Tstr_include incl, sg, new_env
-    | Pstr_extension (ext, _attrs) ->
-        raise (Error_forward (Builtin_attributes.error_of_extension ext))
-    | Pstr_attribute x ->
-        Builtin_attributes.warning_attribute x;
-        Tstr_attribute x, [], env
-  in
   let rec type_struct env sstr =
-    Ctype.init_def(Ident.current_time());
+    Ctype.init_def (Ident.current_time());
     match sstr with
     | [] -> ([], [], env)
-    | pstr :: srem ->
-        let previous_saved_types = Cmt_format.get_saved_types () in
-        let desc, sg, new_env = type_str_item env srem pstr in
-        let str = { str_desc = desc; str_loc = pstr.pstr_loc; str_env = env } in
-        Cmt_format.set_saved_types (Cmt_format.Partial_structure_item str
-                                    :: previous_saved_types);
-        let (str_rem, sig_rem, final_env) = type_struct new_env srem in
-        (str :: str_rem, sg @ sig_rem, final_env)
+    | {pstr_desc = desc; pstr_loc = loc} :: srem ->
+        match desc with
+        | Pstr_eval (sexpr, attrs) ->
+            let expr =
+              Builtin_attributes.warning_scope attrs
+                (fun () -> Typecore.type_expression env sexpr)
+            in
+            let str_rem, sig_rem, final_env = type_struct env srem in
+            mkstr (Tstr_eval (expr, attrs)) env loc :: str_rem,
```

Comment:
Previously, set_saved_types was called before recursing on the rest of the list.  I think this was important, so that we get the Partial_structure_item even if something fails later in the structure.

Reply:
(Reply to comment below as well https://github.com/ocaml/ocaml/pull/1506#discussion_r177030035)

1. In this patch I added a new `mkstr` function which is similar to `mksig`. However, neither `mkstr` in `type_struct` (my patch)  nor `mksig` in `transl_signature` (current trunk) is invoked in correct order. Recursive call goes first and then `mkstr`/`mksig` is called. This should be fixed.

2. I'd like to have `type_structure` and `transl_signature` use same kind of "loop". This change indeed breaks the "simple loop structure" as you said in another comment, but now `type_structure` is looping like `transl_signature`. I think initially I did what you suggested in https://github.com/ocaml/ocaml/pull/1506#discussion_r177030035 but special handling looks ugly...

What do you think?